src/common/sendpacket.c

```diff
@@ -1333,3 +1381,159 @@ sendpacket_abort(sendpacket_t *sp)
 
     sp->abort = true;
 }
+#ifdef HAVE_LIBXDP
+static struct xsk_socket_info *xsk_configure_socket(struct xsk_umem_info *umem, struct xsk_socket_config* cfg, int queue_id, const char *device)
+{
+	struct xsk_socket_info *xsk;
+	struct xsk_ring_cons *rxr = NULL;
+	struct xsk_ring_prod *txr;
+	int ret;
+	xsk = (struct xsk_socket_info*)safe_malloc(sizeof(struct xsk_socket_info));
+	xsk->umem = umem;
+	ret = xsk_socket__create(&xsk->xsk, device, queue_id, umem->umem, rxr, &xsk->tx, cfg);
+	if (ret){
+        return NULL;
+    }
+
+	xsk->app_stats.rx_empty_polls = 0;
+	xsk->app_stats.fill_fail_polls = 0;
+	xsk->app_stats.copy_tx_sendtos = 0;
+	xsk->app_stats.tx_wakeup_sendtos = 0;
+	xsk->app_stats.opt_polls = 0;
+	xsk->app_stats.prev_rx_empty_polls = 0;
+	xsk->app_stats.prev_fill_fail_polls = 0;
+	xsk->app_stats.prev_copy_tx_sendtos = 0;
+	xsk->app_stats.prev_tx_wakeup_sendtos = 0;
+	xsk->app_stats.prev_opt_polls = 0;
+
+	return xsk;
+}
+
+static sendpacket_t * sendpacket_open_xdp(const char *device, char *errbuf){
+    sendpacket_t *sp;
+
+    assert(device);
+    assert(errbuf);
+
+    int nb_of_frames = 131072; // 2^17
```

Comment:
I copied these constants from the example program. I don't really know how to set them. As far as I remember there is only one requirement, memory size should be a power of 2.

Reply:
I think frame_size should be equal to MTU. 

nb_of_frames (umem), nb_of_completion_queue_desc, nb_of_fill_queue_desc should be equal to batch_size due to batch processing

Reply:
Keep in mind libxdp have its defaults too: https://github.com/xdp-project/xdp-tools/blob/master/headers/xdp/xsk.h#L191
If you dont set it explicitly, it will use those (see xsk.c).

Power of 2 umem and frame size is a safe default, the important thing on older kernels that the umem should be memory aligned correctly. Alignment means its starting address is multiply of PAGE_SIZE (4096 on x86).

Newer kernels relaxed these requirements, for example they support unaligned mode:
https://github.com/xdp-project/xdp-tools/tree/master/lib/libxdp#data-path

But if we wont want to deal with checking kernel versions and alignment modes, its safe to stick with these